graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -0,0 +1,183 @@
+
+    The output AST contains a single query type, whose name is the input string. The query type
+    is guaranteed to be the second entry of Document definitions, after the schema definition.
+    The query type has no fields.
+
+    Args:
+        query_type: str, name of the query type for the schema
+
+    Returns:
+        Document, representing a nearly blank schema
+    """
+    blank_ast = ast_types.Document(
+        definitions=[
+            ast_types.SchemaDefinition(
+                operation_types=[
+                    ast_types.OperationTypeDefinition(
+                        operation='query',
+                        type=ast_types.NamedType(
+                            name=ast_types.Name(value=query_type)
+                        ),
+                    )
+                ],
+                directives=[],
+            ),
+            ast_types.ObjectTypeDefinition(
+                name=ast_types.Name(value=query_type),
+                fields=[],
+                interfaces=[],
+                directives=[],
+            ),
+        ]
+    )
+    return blank_ast
+
+
+def merge_schemas(schemas_dict):
+    """Check that input schemas do not contain conflicting definitions, then merge.
+
+    Merged schema will contain all type, interface, enum, scalar, and directive definitions
+    from input schemas. The fields of its query type will be the union of the fields of the
+    query types of each input schema.
+    schemas.
+
+    Args:
+        schemas_dict: OrderedDict where keys are schema_identifiers, and values are
+                      ASTs (Documents) describing schemas. The ASTs will not be modified by
+                      this function
+
+    Returns:
+        MergedSchemaDescriptor, a namedtuple that contains the AST of the merged schema,
+        and the map from names of types/query type fields to the id of the schema that they
+        came from
+
+    Raises:
+        SchemaStructureError if the schema does not have the expected form; in particular, if
+        the AST does not represent a valid schema, if any query type field does not have the
+        same name as the type that it queries, if the schema contains type extensions or
+        input object definitions, or if the schema contains mutations or subscriptions
+        SchemaNameConflictError if there are conflicts between the names of
+        types/interfaces/enums/scalars, or conflicts between the definition of directives
+        with the same name
+    """
+    if len(schemas_dict) == 0:
+        raise ValueError(u'Expected a nonzero number of schemas to merge.')
+
+    query_type = 'RootSchemaQuery'
+    # NOTE: currently, the query type will always be named RootSchemaQuery
+    # could be changed so the user has an input, or by changed to always use the query type
+    # name in the first schema in the input, if desired
+    merged_schema_ast = basic_schema_ast(query_type)  # Document
```

Comment:
Not a huge fan of having three variables all point to parts of the same underlying python object.

Reply:
I think that this makes the mutations to the AST a bit too implicit.